aie_kernels/aie2/reduce_add.cc

```diff
@@ -13,35 +13,32 @@
 #include <stdlib.h>
 #include <type_traits>
 
-#include "../aie_kernel_utils.h"
 #include <aie_api/aie.hpp>
 
-static void _reduce_add_scalar(int32_t *restrict in, int32_t *restrict out,
-                               const int32_t input_size) {
+template <typename T_in, typename T_out, const int N>
+void reduce_add_scalar_int32(T_in *a, T_out *c) {
   event0();
-  int32_t running_total = 0;
-  for (int32_t i = 0; i < input_size; i++) {
-    running_total = running_total + in[i];
+  int32_t sum=0;
+  for (int i = 0; i < N; i++) {
+    sum = sum + a[i];
   }
-  *out = running_total;
+  *(int32_t *)c = sum;
   event1();
-  return;
 }
 
-static void _reduce_add_vector(int32_t *restrict in, int32_t *restrict out,
-                               const int32_t input_size) {
+template <typename T_in, typename T_out, const int N>
+void reduce_add_vector_int32(T_in *in, T_out *out) {
   event0();
   v16int32 zero = broadcast_to_v16int32((int32_t)0);
   const int32_t vector_size = 16;
   v16int32 after_vector;
   v16int32 running_total = zero;
-  AIE_PREPARE_FOR_PIPELINING
-  AIE_LOOP_MIN_ITERATION_COUNT(8)
```

Comment:
This appears to revert the changes from https://github.com/Xilinx/mlir-aie/pull/2248